### PR TITLE
Add doppler secrets edit command

### PR DESCRIPTION
**Pull request by apazzolini**
This works pretty well from my testing, but there are more edge cases to handle before this is ready to merge. At a minimum, this needs:

- [ ] Tests around the parsing logic
- [ ] Tests around edge cases (like saving an empty file or a file that we can't parse)
- [ ] Delete secret support? (compare against the ones you chose, delete any not present in the edited version)
- [ ] Better copy and help text
- [ ] Research if Windows has a way to edit a file that's in memory and not on disk

Demo workflow:

https://user-images.githubusercontent.com/869471/135178810-502ed8c8-8a85-47b3-b243-a5db19d13e96.mov

### CODE REVIEW

File: pkg/cmd/secrets.go [hunk 1]
```diff
@@ -553,5 +562,7 @@ func init() {
 	secretsSubstituteCmd.Flags().String("output", "", "path to the output file. by default the rendered text will be written to stdout.")
 	secretsCmd.AddCommand(secretsSubstituteCmd)
 
+	secretsCmd.AddCommand(secretsEditCmd)
```
Comment (Piccirello) [inline pkg/cmd/secrets.go:565]: We'll want to add the `project` and `config` flags for consistency with our other commands.

File: pkg/cmd/secrets.go [hunk 2]
```diff
@@ -139,6 +139,15 @@ JSON Secret: "{\"logging\": \"info\"}"`,
...
 
+var secretsEditCmd = &cobra.Command{
+	Use:     "edit",
+	Short:   "Edit secrets in your editor",
+	Long:    "Shows a list of current secrets in the config and allows you to edit in your $EDITOR",
+	Example: `$ doppler secrets edit`,
+	Args:    cobra.NoArgs,
+	Run:     editSecrets,
+}
+
```
Comment (Piccirello) [inline pkg/cmd/secrets.go:150]: It would be really cool if this showed a summary of the changes after exiting the viewer, like what secrets were modified, and allowed you to either confirm, cancel, or continue editing. This could also be a future enhancement.

File: pkg/cmd/secrets_edit.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	"strings"
+
+	"github.com/DopplerHQ/cli/pkg/configuration"
+	"github.com/DopplerHQ/cli/pkg/http"
+	"github.com/DopplerHQ/cli/pkg/models"
+	"github.com/DopplerHQ/cli/pkg/utils"
+	"github.com/spf13/cobra"
+)
+
+var header = `# Save this file and quit your editor to save all secrets below.
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: nit: this can be const

File: pkg/cmd/secrets_edit.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	modifiedSecrets := EditInEditor(buffer)
+
+	if buffer == modifiedSecrets {
+		utils.HandleError(fmt.Errorf("File was unchanged, not writing secrets."))
+	}
+
+	secretsToSave := parseEditFormat(modifiedSecrets)
+	saveSecrets(localConfig, secretsToSave)
+
+	fmt.Print("Secrets have been updated.\n")
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: nit: can we use `fmt.Println` and remove the newline character?

File: pkg/cmd/secrets_edit.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+			value = strings.TrimSuffix(secretArr[1], "\n")
+		}
+		secretsMap[name] = value
+	}
+
+	return secretsMap
+}
+
+func isEditableSecret(name string) bool {
+	if name == "DOPPLER_CONFIG" || name == "DOPPLER_ENVIRONMENT" || name == "DOPPLER_PROJECT" {
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: We have a couple other legacy managed secrets we'll want to include: `["DOPPLER_ENCLAVE_PROJECT", "DOPPLER_ENCLAVE_ENVIRONMENT", "DOPPLER_ENCLAVE_CONFIG"]`

File: pkg/cmd/secrets_edit.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		utils.HandleError(httpErr.Unwrap(), httpErr.Message)
+	}
+}
+
+func mapToEditFormat(selectedSecrets []string, secrets map[string]models.ComputedSecret) string {
+	buffer := header
+
+	for idx, val := range selectedSecrets {
+		buffer = buffer + val
+		buffer = buffer + "\n" + secrets[val].RawValue
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: nit: it's more idiomatic to use `fmt.Sprintf` here

```go
buffer = fmt.Sprintf("%s%s\n%s", buffer, val, ecrets[val].RawValue)
```

File: pkg/cmd/secrets_edit.go [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+		utils.HandleError(err, "Unable to launch editor")
+	}
+
+	err = cmd.Wait()
+	if err != nil {
+		utils.HandleError(err, "Unable to wait for editor")
+	}
+
+	content, _ := os.ReadFile(tmpFilePath)
+	defer unix.Close(tmpFd)
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: Since this call is deferred, can we move it to immediately after the fd is returned by `utils.Memfile`? That way an intermediate error doesn't prevent us from closing it.

File: pkg/utils/io.go [hunk 8]
```diff
@@ -76,3 +77,29 @@ func WriteTempFile(name string, data []byte, perm os.FileMode) (string, error) {
 
 	return tmpFileName, nil
 }
+
+func Memfile(name string, b []byte) (int, error) {
+	fd, err := unix.MemfdCreate(name, 0)
+	if err != nil {
+		return 0, fmt.Errorf("MemfdCreate: %v", err)
+	}
```
Comment (Piccirello) [inline pkg/utils/io.go]: Is it guaranteed that if this function returns an error it hasn't created the file descriptor? Otherwise we should close it in the error handler.

File: pkg/cmd/secrets_edit.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+	env := os.Environ()
+	editor := "vim"
+	for _, envVar := range env {
+		// key=value format
+		parts := strings.SplitN(envVar, "=", 2)
+		key := parts[0]
+		if strings.ToUpper(key) == "EDITOR" {
+			editor = parts[1]
+		}
+	}
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: Since this surreptitiously relies on the environment, we may want to introduce out somewhat-standard `no-read-env` flag here.

File: pkg/cmd/secrets_edit.go [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+	tmpFd, err := utils.Memfile("fileToEdit", []byte(contents))
+	if err != nil {
+		utils.HandleError(err, "Unable to create file descriptor")
+	}
+
+	// filepath to our newly created in-memory file descriptor
+	tmpFilePath := fmt.Sprintf("/proc/self/fd/%d", tmpFd)
+
+	env := os.Environ()
+	editor := "vim"
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: Is there any risk of vim writing a backup file, swap file, or some other file to disc? I do see your `noswapfile` flag below.

We'll probably want to document the risk of using other text editors, maybe in the description of the command.

File: pkg/cmd/secrets_edit.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+func EditInEditor(contents string) string {
+	tmpFd, err := utils.Memfile("fileToEdit", []byte(contents))
+	if err != nil {
+		utils.HandleError(err, "Unable to create file descriptor")
+	}
+
+	// filepath to our newly created in-memory file descriptor
+	tmpFilePath := fmt.Sprintf("/proc/self/fd/%d", tmpFd)
```
Comment (Piccirello) [inline pkg/cmd/secrets_edit.go]: What permissions does this fd exist with? Is there any risk of other processes accessing this file? I'm wondering if we can lock it down more, otherwise this may be some fd that any other process (or at least any other process running as this user) can access. In which case it's no different than just writing it to the filesystem.

File: pkg/utils/io.go [hunk 12]
```diff
@@ -76,3 +77,29 @@ func WriteTempFile(name string, data []byte, perm os.FileMode) (string, error) {
...
+	if err != nil {
+		return 0, fmt.Errorf("Mmap: %v", err)
+	}
+
+	copy(data, b)
+
+	err = unix.Munmap(data)
+	if err != nil {
+		return 0, fmt.Errorf("Munmap: %v", err)
+	}
```
Comment (Piccirello) [inline pkg/utils/io.go]: I don't fully understand this pattern, in part because Go doesn't provide any docs on this command, and linux's [mmap](https://www.man7.org/linux/man-pages/man2/mmap.2.html) doesn't take an fd. How is that we can copy to `data` and then immediately unmap it from memory?

File: pkg/cmd/secrets.go [hunk 13]
```diff
@@ -139,6 +139,15 @@ JSON Secret: "{\"logging\": \"info\"}"`,
...
 }
 
+var secretsEditCmd = &cobra.Command{
+	Use:     "edit",
+	Short:   "Edit secrets in your editor",
+	Long:    "Shows a list of current secrets in the config and allows you to edit in your $EDITOR",
+	Example: `$ doppler secrets edit`,
+	Args:    cobra.NoArgs,
+	Run:     editSecrets,
+}
```
Comment (Piccirello) [inline pkg/cmd/secrets.go:149]: Does this gracefully handle secret renames? If not we should just add a note about that limitation to the help text.

**Review by Piccirello — COMMENTED** [review]
This is really f'in cool

File: pkg/utils/io.go (hunk 12 above)
Comment (apazzolini) [reply pkg/utils/io.go]: Great question. The snippet came from [this blog post](https://terinstock.com/post/2018/10/memfd_create-Temporary-in-memory-files-with-Go-and-Linux/) via @zbuc. I also don't fully understand the pattern, but it does indeed work..

File: pkg/cmd/secrets_edit.go (hunk 11 above)
Comment (apazzolini) [reply pkg/cmd/secrets_edit.go]: I tested, and other processes running as the same user can still access the file. Chris was going to research the flags to see if we can lock it down further, but I'm not sure if he's had a chance to do so yet. I think it's still better than writing to disk though.

Comment (apazzolini) [reply pkg/cmd/secrets_edit.go]: Discussed offline 👍

File: pkg/cmd/secrets_edit_linux_amd64.go [hunk 14]
```diff
@@ -0,0 +1,226 @@
+/*
+Copyright © 2020 Doppler <support@doppler.com>
```
Comment (ScottBrenner) [inline pkg/cmd/secrets_edit_linux_amd64.go:2]:
```suggestion
Copyright © 2022 Doppler <support@doppler.com>
```

**Comment by apazzolini** [comment]
Closing - we won't be doing this.
